### PR TITLE
Add missing migrations

### LINKED ISSUE #1542: Add missing ledger db migrations
## Summary
We have some name changes on ledger db tables after we implemented light client proofs, we added namespaces to those tables like `...BatchProofs` or `...LightClientProofs`. In order to be able to sync from the beginning on the current nightly version we need to implement necessary migrations

### PR DESCRIPTION
**Pull request by ercecan**
# Description
Adds migration of ledger db table name change
VerifiedProofsBySlotNumber -> VeriifiedBatchProofsBySlotNumber

Also adds an api to help make this easier for the future migrations

## Linked Issues
- Fixes #1542

### CODE REVIEW

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs
Comment (rakanalh) [inline crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs]:
```suggestion
pub struct UnsignedSoftConfirmationV1<'txs> {
```

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs
Comment (rakanalh) [inline crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs]: This can be in an 
```
impl Into<UnsignedSoftConfirmation> for UnsignedSoftConfirmationV1 
```

File: crates/sovereign-sdk/module-system/sov-modules-stf-blueprint/src/lib.rs
Comment (rakanalh) [inline crates/sovereign-sdk/module-system/sov-modules-stf-blueprint/src/lib.rs]:
```suggestion
    let message = borsh::to_vec(&unsigned_soft_confirmation.into()).unwrap();
```

File: .gitignore
Comment (rakanalh) [inline .gitignore]: Anything we write to disk should go into `resources`. We should keep the root clean of dynamically created files / directories.

Comment (rakanalh) [reply .gitignore]: I also searched where we use this directory and could not find any usage of this.

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (rakanalh) [inline crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: I believe column families are rocksdb related. Please include them in `RocksdbConfig`

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [inline crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: This can be moved outside `LedgerDbMigrator` above `clear_db_dir`

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (rakanalh) [inline crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: These don't belong to `LedgerDB`... maybe add them as migration helpers / utils?

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [inline crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: I don't believe that dropping tables is the responsibility of the migration manager. Each migration should decide what it wants to do whether it's to migrate old tables to new or remove tables .. etc. Please call `drop_column_family` inside the migration itself.

**Review by rakanalh — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (ercecan) [reply .gitignore]: Yeah my bad, it is something else will remove

File: crates/fullnode/src/runner.rs
Comment (rakanalh) [inline crates/fullnode/src/runner.rs]: This could have belonged to it's own PR since it's unrelated to the main context of the PR.

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: I cannot do that because it is still locked outside of the migration itself
I agree that it is not the responsibility but the migration it self decides whether to drop it or not by adding them to a vector.
To do it as you say I will need to:
1-  pass the temp db path to the execute function
2- create ledger db
3- Do the migration stuff
4- Drop temp db
5- Drop cf
6- Repeat this process on every single migration

File: crates/fullnode/src/runner.rs
Comment (ercecan) [reply crates/fullnode/src/runner.rs]: Yeah, did not want to bother tbh :D

Comment (ercecan) [reply crates/fullnode/src/runner.rs]: Should I?

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs
Comment (ercecan) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs]: These changes are cherry-picked from #1514 
I believe I should wait for him to change these parts

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs
Comment (ercecan) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs]: These changes are cherry-picked from #1514 
I believe I should wait for him to change these parts

File: crates/fullnode/src/runner.rs
Comment (rakanalh) [reply crates/fullnode/src/runner.rs]: yes please

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: I don't understand that.

- The migration `execute` receives an instance of the `new_ledger_db` which is already using the one in tempdir.
- Do step number 3
- Drop cf

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs
Comment (rakanalh) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs]: Commented there on this.

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: On second thought, I am not sure if it should be in the config, because we can already get it with a function just by giving the path

Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: I'll think on this a bit more

Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: It also kinda makes sense as well :D

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: I cannot drop cf without dropping all instances of a rocksdb

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs:162]: How about I put them in LedgerDBMigrator instead of utils or sth like that?

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: What happens if you don't drop it?

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: Okay, so I think let's skip that one for now because the column_families should be optional and if it is None it should be either LEDGER_TABLES STATE_TABLES or NATIVE_STATE_TABLES, to extract which to use we can use the path or pass another parameter but I feel like that's just extra code

I recommend implementing this if we really need to

WDYT?

File: crates/fullnode/src/runner.rs
Comment (ercecan) [reply crates/fullnode/src/runner.rs]: #1560

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: 1) The ledger db will not open with the usual LEDGER_TABLES and cry missing column family etc etc because the struct is removed from LEDGER_TABLES
2) It stays there and occupies extra space

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: I am not sure i follow. In either way, we are passing `column_families` here... why not move this param to `RocksdbConfig` and use it the exact same way you do in the PR?

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: Okay. Let me take a stab at it tomorrow morning and i'll let you know the result.

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs
Comment (ercecan) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: I can do that for ledger db but for example here I will need to create two configs just for that, there are other examples to this as well, I think this will just generate more code + all of the column families are accessible publicly by all crates importing sov-db as they are pub consts
Screenshot 2024-12-05 at 22 48 48

Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs]: In this example, we pass `None` and then `unwrap_or_else` when the default list of CFs are needed.

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs]: > I cannot drop cf without dropping all instances of a rocksdb

I see your point now. The problem is that `DB` used by ledger is behind an `Arc` and `drop_cf` needs to borrow a mutable reference. Which is why we need to call `drop_cf` with the instance of `mut DB`

**Comment by rakanalh** [comment]
Should be merged after: https://github.com/chainwayxyz/citrea/pull/1514

To fix the cherry-picked stuff from the merged version.

**Review by rakanalh — APPROVED** [review]

**Review by eyusufatik — APPROVED** [review]
after merging #1514 you need to remove OldFormat stuff and we are good to go

**Comment by eyusufatik** [comment]
I reverted

**Pull request merged**
